### ISSUE TITLE
geth js exits after startup on unstable

### ISSUE THREAD

**Issue by kevinday** [ethereum/go-ethereum#2893]
```
Geth
Version: 1.5.0-unstable
Protocol Versions: [63 62]
Network Id: 1
Go Version: go1.5.1
OS: linux
GOPATH=
GOROOT=/usr/lib/go
```
When running geth 1.5.0-unstable with the js option, it starts up then shuts down immediately. I have a `test.js` file that just contains `console.log("test");`

On a clean VM, I'm doing
```
sudo apt-get -y install software-properties-common
sudo add-apt-repository -y ppa:ethereum/ethereum
sudo add-apt-repository -y ppa:ethereum/ethereum-dev
sudo apt-get update
sudo apt-get -y install ethereum
geth js test.js
```
and I see:
```
I0807 05:57:13.073414 ethdb/database.go:82] Allotted 128MB cache and 1024 file handles to /home/auguruser/.ethereum/chaindata
I0807 05:57:13.424525 cmd/utils/flags.go:849] ------------------------------------------------------------------------------------------------
I0807 05:57:13.424653 cmd/utils/flags.go:850] Geth is currently configured to SUPPORT the DAO hard-fork!
I0807 05:57:13.424696 cmd/utils/flags.go:851] You can change your choice prior to block #1920000 with --support-dao-fork or --oppose-dao-fork.
I0807 05:57:13.424750 cmd/utils/flags.go:852] After the hard-fork block #1920000 passed, changing chains requires a resync from scratch!
I0807 05:57:13.424792 cmd/utils/flags.go:853] ------------------------------------------------------------------------------------------------
I0807 05:57:13.424932 ethdb/database.go:169] closed db:/home/auguruser/.ethereum/chaindata
I0807 05:57:13.425034 cmd/utils/flags.go:592] WARNING: No etherbase set and no accounts found as default
I0807 05:57:13.425697 cmd/geth/main.go:309] instance: Geth/1.5.0-unstable/go1.5.1/linux
I0807 05:57:13.425747 ethdb/database.go:82] Allotted 128MB cache and 1024 file handles to /home/auguruser/.ethereum/chaindata
I0807 05:57:13.473078 ethdb/database.go:82] Allotted 16MB cache and 16 file handles to /home/auguruser/.ethereum/dapp
I0807 05:57:13.481495 eth/backend.go:184] Protocol Versions: [63 62], Network Id: 1
I0807 05:57:13.482359 core/blockchain.go:204] Last header: #0 [d4e56740…] TD=17179869184
I0807 05:57:13.482424 core/blockchain.go:205] Last block: #0 [d4e56740…] TD=17179869184
I0807 05:57:13.482484 core/blockchain.go:206] Fast block: #0 [d4e56740…] TD=17179869184
I0807 05:57:13.483748 p2p/server.go:322] Starting Server
I0807 05:57:15.628811 p2p/discover/udp.go:217] Listening, enode://113e6afd1997fd9ba067ef46174a3d53689ebfc46fe2ec4171fae6377948b224a8d2ba04d54cfe91856492eadf7c2f4f696893e88fe4bed28270edf58552746e@[::]:30303
I0807 05:57:15.634179 p2p/server.go:576] Listening on [::]:30303
I0807 05:57:15.637565 node/node.go:296] IPC endpoint opened: /home/auguruser/.ethereum/geth.ipc
test
I0807 05:57:15.908008 node/node.go:328] IPC endpoint closed: /home/auguruser/.ethereum/geth.ipc
I0807 05:57:15.908032 core/blockchain.go:572] Chain manager stopped
I0807 05:57:15.908040 eth/handler.go:224] Stopping ethereum protocol handler...
I0807 05:57:15.908053 eth/handler.go:245] Ethereum protocol handler stopped
I0807 05:57:15.908068 core/tx_pool.go:163] Transaction pool stopped
I0807 05:57:15.908084 eth/backend.go:502] Automatic pregeneration of ethash DAG OFF (ethash dir: /home/auguruser/.ethash)
I0807 05:57:15.908286 ethdb/database.go:169] closed db:/home/auguruser/.ethereum/chaindata
I0807 05:57:15.908330 ethdb/database.go:169] closed db:/home/auguruser/.ethereum/dapp
```
On other versions of geth using `js` does not shut down geth.

**Comment by Hellsegga** [comment]
I took the liberty to investigate this as a way to learn geth and hope I got it right.
Essentially the behavior of geth has not changed: it wants to close down after the execution of the JS file but waits to honor any pending timers (i.e. when setTimeOut or setInterval is called by the JS).
Now there are no timers in your test JS file, but the web3.js is executed before it and this is where the change has been introduced: in previous versions geth bundled an earlier version of web3.js that did set up timers on initialization (in RequestManager.setProvider) whereas is more recent versions timers are no longer setup.
So no timers means geth exits after the JS file has been executed.

**Comment by kevinday** [comment]
Thanks for looking into this Hellsegga. So are you suggesting the intended behavior of `js` has always been to shut down after execution of the js file, and that it was just a coincidence caused by another piece that it wasn't?

I know the js I included in this example is trivial, but myself and others rely on `js` with non trivial pieces of code, like to toggle mining on/off when there are pending transactions. Even if this change is fixing some previously unintended behavior, a lot of us have come to rely on that behavior. It would be great to get some clarification from the devs here about what the intent of `js` is.

**Comment by Hellsegga** [comment]
> So are you suggesting the intended behavior of js has always been to shut down after execution of the js file, and that it was just a coincidence caused by another piece that it wasn't?

Yes that's how I interpret the code

I will let the devs answer on the intended behavior and whether it should be changed, but maybe one of these options would work for you (I understand your use case as wanting to start a node, execute a JS file and then let the node continue to run, did I get it right?):
- `geth --preload test.js console` (will execute your script and then start an interactive console, so will not close the node)
- start your geth node in the background and then run `geth --exec "$(< test.js)" attach` (will create a console that attaches to the node running in the background and execute your file. You then let the node continue to run in the background)

**Comment by kevinday** [comment]
Yeah, I'm running geth in an upstart service with some js loaded to toggle mining on off as transactions appear. So any solution that uses `attach` or `console` won't work because those don't background well.  This has worked fine for me for months, but it looks like this will be breaking soon.

I brought up allowing `--preload` without `console` or `attach` a while ago (https://github.com/ethereum/go-ethereum/pull/2431#issuecomment-212590842), but was told that's what `js` is for so I switched to it instead.

https://github.com/ethereum/go-ethereum/pull/2431#issuecomment-212590842

**Comment by Hellsegga** [comment]
OK. As I understand the current intended behavior it will exit once the javascript doesn't have any timers left (js setTimeOuts in your code, web3 watches, basically anything left to expect from the outer world...). So exiting after the execution of the test file that just does `console.log('test')` is correct but if your code watches for new transactions then I think it should not have exited.
Do you mind sharing the code you use (or at least a mock showing the web3 calls)?

**Comment by fjl** [comment]
Thanks @Hellsegga. I can confirm that this is indeed the correct behaviour. If you have web3.js filters installed, your script will not exit until you uninstall them.

**Comment by kevinday** [comment]
@fjl @Hellsegga 
A bit more info: here is my js:
```
var mining_threads = 1;

function checkWork() {
    if (eth.getBlock("pending").transactions.length > 0) {
        if (eth.mining) return;
        console.log("== Pending transactions! Mining...");
        miner.start(mining_threads);
    } else {
        miner.stop();
        console.log("== No transactions! Mining stopped.");
    }
}

eth.filter("latest", function(err, block) { checkWork(); });
eth.filter("pending", function(err, block) { checkWork(); });

checkWork();
```

I do `geth init` with the following genesis.json:
```
{
  "nonce": "0x0000000000000021",
  "difficulty": "0x00",
  "config": {
    "homesteadBlock": 0,
    "daoForkBlock" : 0,
    "daoForkSupport" : true
  },
  "alloc": {
    "7fbe93bc104ac4bcae5d643fd3747e1866f1ece4": {
      "balance": "1000000000000000000000000000000"
    }
  },
  "mixhash": "0x0000000000000000000000000000000000000000000000000000000000000000",
  "coinbase": "0x0000000000000000000000000000000000000000",
  "timestamp": "0x00",
  "parentHash": "0x0000000000000000000000000000000000000000000000000000000000000000",
  "extraData": "0x11bbe8db4e347b4e8c937c1c8370e4b5ed33adb3db69cbdb7a38e1e50b1b82fa",
  "gasLimit": "0x2fefd8"
}
```

and then run `geth js mining_toggle.js`

geth shuts down immediately (though it doesn't on stable).

**Comment by fjl** [comment]
OK, that should indeed work.

**Comment by fjl** [comment]
Fixed by #2910
